### ISSUE TITLE
Istioctl analyzer:  Clarify IST0102 ( NamespaceNotInjected) for ambient

### ISSUE THREAD

**Issue by birkland** [istio/istio#48103]
IST0102 is a little misleading and confusing  when emitted by `istioctl analyze` on an ambient-capable mesh.

The root cause of the confusion is how the label `istio.io/dataplane-mode=ambient` interacts with this analysis code

In effect, `istio.io/dataplane-mode=ambient` is considered a valid "injection" label that suppresses this warning, despite (a) not really fitting the term "injection", and (b) not being mentioned in the analysis code text as a possible resolution.

**Describe the feature request**

Clarify the text of `IST0102` to accommodate the presence ambient dataplane mode.  `NamespaceNotInjected` is not really an appropriate description now that ambient dataplane mode is available.  Consider more generic terms, like `NamespaceNotLabeledForMesh`.  Re-word the description to include _both_ injection _and_ ambient dataplane mode as possible remedies, as appropriate.

**Describe alternatives you've considered**

Deprecate IST0102 and use a new, more generic message that is applicable to both the injected and ambient worlds.

**Affected product area (please put an X in all that apply)**

`istioctl analyze` and associated docs, when used with an ambient-capable mesh

[ x ] Ambient
[ x ] Docs
[ ] Dual Stack
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ x ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by hanxiaop** [comment]
Which istioctl version arer you using? IIRC we do have some specific logic for this message.

**Comment by ilrudie** [comment]
I can try to repro from latest istioctl and confirm @hanxiaop

**Comment by ilrudie** [comment]
341ae818074e56217b94f2b23f991c91cce133ab still makes no mention of ambient as a possible solution for what's really more of an "not Istio enabled namespace now" instead of a NotInjected namespace

Long term this should probably remove the "injected" verbiage but shorter term potentially we could add to the message a note about ambient only when ambient is enabled in the cluster.

**Comment by hanxiaop** [comment]
> [341ae81](https://github.com/istio/istio/commit/341ae818074e56217b94f2b23f991c91cce133ab) still makes no mention of ambient as a possible solution for what's really more of an "not Istio enabled namespace now" instead of a NotInjected namespace
> 
> Long term this should probably remove the "injected" verbiage but shorter term potentially we could add to the message a note about ambient only when ambient is enabled in the cluster.

Agreed. I think we'd better to create a new message to describe the label state or a new message to describe only the ambient thing, since the original message was only for injection.

**Comment by ilrudie** [comment]
Thinking a bit more about this, detecting "ambient is available" might be tricky. If you have permissions to get/list `istio-system` then we can check if the ambient controllers are enabled but I suspect most general users wouldn't have this level of permission and wouldn't ever see any alternative "you could try `istio.io/dataplane-mode: ambient`" type of message.

**Comment by ilrudie** [comment]
Current message, for easy reference:

> The namespace is not enabled for Istio injection. Run 'kubectl label namespace %s istio-injection=enabled' to enable it, or 'kubectl label namespace %s istio-injection=disabled' to explicitly mark it as not needing injection.

**Comment by ilrudie** [comment]
> Re-word the description to include both injection and ambient dataplane mode as possible remedies, as appropriate.

Does this make sense when ambient isn't production ready though? Istioctl analyze giving advice to use something that _we_ as the developers haven't said is ready?

**Comment by linsun** [comment]
> Current message, for easy reference:
> 
> > The namespace is not enabled for Istio injection. Run 'kubectl label namespace %s istio-injection=enabled' to enable it, or 'kubectl label namespace %s istio-injection=disabled' to explicitly mark it as not needing injection.

Yep so we could modify to "The ns is not enable for Istio.... label istio-injection or dataplane-mode=ambient if ambient is enabled to enable the namespace for Istio."

**Comment by hanxiaop** [comment]
> > Current message, for easy reference:
> > > The namespace is not enabled for Istio injection. Run 'kubectl label namespace %s istio-injection=enabled' to enable it, or 'kubectl label namespace %s istio-injection=disabled' to explicitly mark it as not needing injection.
> 
> Yep so we could modify to "The ns is not enable for Istio.... label istio-injection or dataplane-mode=ambient if ambient is enabled to enable the namespace for Istio."

So, regarding ambient, are we encouraging people to use mixed modes? I think this message might give users the impression that both modes are acceptable. My original intention was to avoid mentioning sidecar injection in messages related to ambient, so pure ambient users won't see anything related to the sidecar. I think it would be better to prevent ambient logic from causing false positives in the current message, and then generate a new ambient message that doesn't mention sidecar injection once the label is mature?

**Comment by ilrudie** [comment]
@hanxiaop, AFAIK interop will at some point be a priority although potentially not for initial ambient Beta. At any rate I do agree that we need to be careful with wording at least until we deem ambient+sidecar interop stable to where Istio as a project feels safe with mixed use cases in production.

**Comment by ilrudie** [comment]
Actually interesting side note here: at present `istio.io/dataplane-mode=ambient` satisfies `istioctl analyze` regardless of whether ambient controllers are running (makes sense, it's not trivial to detect). In no way are you meaningfully "enrolled" in the mesh but there's no alert.

In practice it's probably not a huge issue right now since ambient is still not something we suggest everyone run but this may become a more urgent problem if there's a gap between "ambient controllers always enabled" and "ambient is beta/GA" where folks may be confused depending on what their mesh administrators have done.

**Comment by bleggett** [comment]
- Whether a namespace is properly labeled for mesh inclusion
- Whether your particular cluster and mesh configuration currently has the right number and kind of healthy cluster-wide pods/controllers/operators/etc  

Should probably be disjoint checks, for disjoint roles.

**Comment by ilrudie** [comment]
Don't disagree that they can/should be separate but for safety (even though it'll be pretty annoying to users who don't have ambient available) we probably need to have a check for `istio.io/dataplane-mode=ambient` on your namespace but 1 or more pods has no capture annotation before we think about suggesting that ambient label might fix your enrollment in the mesh problem.

**Comment by ilrudie** [comment]
Having istioctl suggest that folks use ambient while it's still not beta is probably a non-starter. How can we suggest something that almost certainly isn't installed and running? If it is installed and running today you are either aware because you're testing it or you probably shouldn't be using it anyway. It may even be unwise to suggest ambient until the ambient controllers become enabled by default.

A possible remedy that strictly makes ambient more annoying would be to just drop `istio.io/dataplane-mode: ambient` as an "injection label" so that even if you have this label we'll alert folks that they aren't getting sidecars. This probably isn't that valuable though since unless we back port it to previous versions ambient is hopefully released and plausibly available by the time this change makes it to istioctl.

Another would be to implement a new analyzer which checks for a state such as: has `istio.io/dataplane-mode: ambient` && pods lack `ambient.istio.io/redirection: enabled` annotation. Basically, user intends to use ambient but it's not working. This isn't 100% fool proof since there is a possible scenario where ambient was enabled but was for some reason backed out and to join the mesh you need sidecars again. This should eventually work again though because as pods recycle they don't pickup the redirection annotation.

**Comment by ilrudie** [comment]
I don't think `ambient.istio.io/redirection: enabled` is covered by Louis's layered future proposal. If we want to use it as a mechanism to help end user's without elevated mesh privileges understand their WLs' enrollment status we may want to codify this as a requirement for conformance.

**Comment by linsun** [comment]
Is this still an issue? @birkland @ilrudie

**Comment by ilrudie** [comment]
I don't recall this being an issue anymore in gmc when I was testing but that cluster is gone now. Its fixed in istioctl itself now at least.
